design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+json_concat("{\"a\": {\"5\": 6}}", "{\"a\": {\"3\": 4}}") // returns "{\"a\": {\"3\": 4}}"
+json_concat("{}", "{}") // returns "{}"
+json_concat("[1, 2]", "[3, 4]") // returns "[1,2,3,4]"
+json_concat("[1, [2]]", "[[[3]], [[[4]]]]") // returns "[ 1, [2], [[3]], [[[4]]] ]"
+json_concat("null", "null") // returns [null, null]
+json_concat("[1, 2]", "{\"a\": 1}") // returns "[1, 2, {\"a\": 1}]"
+json_concat("[1, 2]", "3") // returns "[1, 2, 3]"
+json_concat("[3]", "[1, 2]") // returns "[3, 1, 2]"
+json_concat("1", "2") // returns "[1, 2]"
+json_concat("[]", "[]") // returns []
+json_concat("abc", "[1]") // returns NULL
+json_concat(NULL, "[1]") // returns NULL
+```
+
+#### Rejected alternatives
+
+If one argument represents a valid JSON and the other one does not, then return the valid string
+as a result of concatenation.
+
+Given that concatenation of invalid JSON strings is an exceptional situation, I believe that we
+should return `NULL` if any input is not a JSON. This way, we avoid ambiguity if a user needs to trace
+back how they got certain results.
+
+If this function returns the valid record in the case when the other one is invalid, from the user
+perspective, there are 3 possibilities how they ended up with this result:
+* The second object could have been empty.
+* The second object could have had the same keys/values.
+* The second object is invalid JSON.
+
+By always returning `NULL` given an invalid JSON string we remove the third, the least obvious,
+possibility which reduces cognitive load from the user.
+
+### to_json_string
+
+```
+to_json_string(val)
+```
+
+Given any ksqlDB type returns the equivalent JSON string.
+
+#### Examples
+
+```
+to_json_string(1) // returns "1"
+to_json_string(15.3) // returns "15.3"
+to_json_string("abc") // returns "\"abc\""
+to_json_string(Array[1, 2, 3]) // returns "[1, 2, 3]"
+to_json_string(Struct{id=1,name=A}) // returns "{\"id\": 1, \"name\": \"a\"}"
+to_json_string(Map('c' := 2, 'd' := 4)) // returns "{\"c\": 2, \"d\": \"4\"}"
+to_json_string(NULL) // returns "null"
+```
+
+#### Rejected alternatives
+
+Return `NULL` given `NULL` as input. Although, it might make sense from the `NULL` propagation point
+of view, there are several arguments why we should map SQL `NULL` to JSON `null`:
+* Akin to SQL's `NULL`, JSON's `null` is a marker of absent value, so semantics does not change.
+* Returning SQL `NULL` would make it impossible to serialize any input to JSON `null`.
+
+## Design
+
+Akin to the existing `EXTRACTJSONFIELD` and `JSON_ARRAY_CONTAINS`, the general workflow for all
+functions is to parse the given string(s) into jackson's `JsonNode`s and perform manipulations according
+to the functions' specifications.
+
+This approach might be suboptimal from the performance point of view since every function invocation
+completely deserializes the entire structure and then serializes the result back. However, I would argue
```

Comment:
nit: remove `I would argue`

Reply:
Removed in https://github.com/confluentinc/ksql/pull/8550/commits/9672f21f932c16bf173529889608035ce467270c.